### PR TITLE
skip columns with variance = 0

### PR DESCRIPTION
**Pull request by jwist**
with this the pca would skip all features with variance = 0 answering the issue https://github.com/mljs/pca/issues/10

I think it is fair to skip these data, but warn user that columns have been skipped

### CODE REVIEW

**Comment by targos** [comment]
The problem is that then you cannot use the predict method without removing that column yourself, or can you?

**Comment by jwist** [comment]
> The problem is that then you cannot use the predict method without removing that column yourself, or can you?

you are right. To solve this I flagged the zero variance column and remove it before the prediction.

**Comment by targos** [comment]
I changed the PR to make it an option (ignoreZeroVariance) and fixed updating the index in _adjust. What do you think?

**Comment by jwist** [comment]
> I changed the PR to make it an option (ignoreZeroVariance) and fixed updating the index in _adjust. What do you think?

Much nicer like this

**Comment by jwist** [comment]
I just replace the useX flags by a proper switch. This is because useCovarianceMatrix=true would inhibit the useNIPALS=true in the cascaded tests. I think using a option.method='svd' as default is clearer. I hope we don't break to much code backwards????

File: pca.d.ts [hunk 1]
```diff
@@ -3,11 +3,11 @@ import { Matrix } from 'ml-matrix';
 declare module 'ml-pca' {
   export interface IPCAOptions {
     isCovarianceMatrix?: boolean;
-    useCovarianceMatrix?: boolean;
+    method?: String;
```
Comment (targos) [inline pca.d.ts]:
```suggestion
    method?: 'SVD' | 'NIPALS' | 'covarianceMatrix';
```

File: src/__tests__/iris.js [hunk 2]
```diff
@@ -25,7 +25,7 @@ const expectedLoadingsNIPALS = [
 ];
 
 describe('iris dataset', function () {
-  var pca = new PCA(iris, { scale: true, useCovarianceMatrix: false });
+  var pca = new PCA(iris, { scale: true, method: 'svd' });
```
Comment (targos) [inline src/__tests__/iris.js]:
```suggestion
  var pca = new PCA(iris, { scale: true, method: 'SVD' });
```

File: src/pca.js [hunk 3]
```diff
@@ -2,14 +2,14 @@ import { Matrix, MatrixTransposeView, EVD, SVD, NIPALS } from 'ml-matrix';
...
+ * @param {Matrix} dataset - dataset or covariance matrix.
  * @param {Object} [options]
- * @param {boolean} [options.isCovarianceMatrix=false] - true if the dataset is a covariance matrix
- * @param {boolean} [options.useCovarianceMatrix=false] - force the use of the covariance matrix instead of singular value decomposition.
- * @param {boolean} [options.useNIPALS=false] - if true, then NIPALS algorithm is used to compute nCompNIPALS first components
- * @param {boolean} [options.nCompNIPALS=2] - number of components to be computed with NIPALS
- * @param {boolean} [options.center=true] - should the data be centered (subtract the mean)
- * @param {boolean} [options.scale=false] - should the data be scaled (divide by the standard deviation)
+ * @param {boolean} [options.isCovarianceMatrix=false] - true if the dataset is a covariance matrix.
+ * @param {boolean} [options.method='svd'] - select which method to use: svd (default), covarianceMatrirx or NIPALS.
```
Comment (targos) [inline src/pca.js]:
```suggestion
 * @param {boolean} [options.method='SVD'] - select which method to use: SVD (default), covarianceMatrirx or NIPALS.
```

File: src/pca.js [hunk 4]
```diff
@@ -22,63 +22,62 @@ export class PCA {
...
+      this.excludedFeatures = model.excludedFeatures;
       return;
     }
 
     dataset = new Matrix(dataset);
 
     const {
       isCovarianceMatrix = false,
-      useNIPALS = false,
+      method = 'svd',
```
Comment (targos) [inline src/pca.js]:
```suggestion
      method = 'SVD',
```

File: src/pca.js [hunk 5]
```diff
@@ -22,63 +22,62 @@ export class PCA {
...
-      this.U = svd.rightSingularVectors;
-
-      const singularValues = svd.diagonal;
-      const eigenvalues = [];
-      for (const singularValue of singularValues) {
-        eigenvalues.push((singularValue * singularValue) / (dataset.rows - 1));
+    this._adjust(dataset, ignoreZeroVariance);
+    switch (method) {
+      case 'covarianceMatrix': {
+      // User provided a dataset but wants us to compute and use the covariance matrix.
```
Comment (targos) [inline src/pca.js]:
```suggestion
        // User provided a dataset but wants us to compute and use the covariance matrix.
```

File: src/pca.js [hunk 6]
```diff
@@ -22,63 +22,62 @@ export class PCA {
...
+        const covarianceMatrix = new MatrixTransposeView(dataset)
+          .mmul(dataset)
+          .div(dataset.rows - 1);
+        this._computeFromCovarianceMatrix(covarianceMatrix);
+      }
+        break;
+      case 'NIPALS':
+        this._computeWithNIPALS(dataset, nCompNIPALS);
+        break;
+      default: {
```
Comment (targos) [inline src/pca.js]:
```suggestion
      case 'SVD': {
```

File: src/pca.js [hunk 7]
```diff
@@ -22,63 +22,62 @@ export class PCA {
...
+
+        this.U = svd.rightSingularVectors;
+
+        const singularValues = svd.diagonal;
+        const eigenvalues = [];
+        for (const singularValue of singularValues) {
+          eigenvalues.push((singularValue * singularValue) / (dataset.rows - 1));
+        }
+        this.S = eigenvalues;
       }
```
Comment (targos) [inline src/pca.js]:
```suggestion
        break;
      }
      default:
        throw new Error(`unknown method: ${method}`);
```

**Comment by targos** [comment]
> I hope we don't break to much code backwards????

Not a problem. I'll make a major release

File: pca.d.ts (hunk 1 above)
Comment (jwist) [reply pca.d.ts]: ok! got it!

**Comment by jwist** [comment]
> > I hope we don't break to much code backwards????
> 
> Not a problem. I'll make a major release

Ok

**Pull request merged**